proposals/shared-everything-threads/Overview.md

```diff
@@ -401,24 +401,26 @@ primitive that does not depend on linear memory. We propose adding a new abstrac
 `waitqueue` to serve as this new primitive.
 
 Unlike other abstract heap types, which are only ever subtypes of other abstract heap types (or are
-bottom types), `waitqueue` is a final subtype of `(rec (type (sub (struct shared (field i32))))).0`,
-meaning it is a struct with one visible `i32` field that can be accessed with all the standard
-struct accessors as well as the new atomic struct accessors. This field is the futex control field
-that is atomically checked when waiting on the waiter queue. `waitqueue` is also always shared.
-There is no non-shared version of it. It is not valid to declare a new subtype of `waitqueue`.
+bottom types), `waitqueue` is a final subtype of `(rec (type (sub (struct shared (field (mut (ref
+null (shared eq)))))))).0`, meaning it is a struct with one visible shared eqref field that can be
```

Comment:
Having something a bit more principled than a random hard-coded type (whichever one it is) would really be nice. What would be the problem with `(waitqueue valtype)`, with some suitable constrains on the valtype, e.g., it must be sharable?

Reply:
I think the right constraint would be `i32, i64, or <: eqref`

Reply:
As a minimal initial design, the instruction could hardcode `i32` as the parameter while leaving wiggle-room in the instruction encoding for a more general type in future.

Reply:
it has to be a type for which we can define atomic `cmpxchg`, which doesn't currently include (e.g.) `f32` and `f64`, or any non-`eq` ref.

Reply:
I see, thanks for the clarification.

Reply:
I'll add a note about the possibility of parameterization for now. I agree that it would be nice to have, but we can stick with hard coded i32 until we have enough implementer feedback to know we're not shooting off in the wrong direction.